### PR TITLE
[R-package] Add GPU install options (fixes #3765)

### LINKED ISSUE #3765: R package install with GPU support fails
This used to work:

```
FROM nvidia/cuda:11.0-devel-ubuntu20.04

RUN apt-get update && \
    DEBIAN_FRONTEND="noninteractive" apt-get install -y software-properties-common apt-transport-https

RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9 && \
    add-apt-repository 'deb [arch=amd64] https://cran.rstudio.com/bin/linux/ubuntu focal-cran40/' && \
    apt-get update && \
    apt-get install -y r-base

RUN apt-get install -y git wget libcurl4-openssl-dev default-jdk-headless libssl-dev libxml2-dev cmake

ENV MAKE="make -j$(nproc)"

RUN R -e 'install.packages(c("R6","data.table","jsonlite"), repos = "https://cran.rstudio.com/")'

RUN apt-get install -y libboost-dev libboost-system-dev libboost-filesystem-dev ocl-icd-opencl-dev opencl-headers clinfo

RUN mkdir -p /etc/OpenCL/vendors && \
    echo "libnvidia-opencl.so.1" > /etc/OpenCL/vendors/nvidia.icd   ## otherwise lightgm segfaults at runtime (compiles fine without it)

RUN git clone --recursive https://github.com/microsoft/LightGBM && \
    cd LightGBM && \
    Rscript build_r.R --use-gpu
```

Now, I get this error:

```
Cloning into 'LightGBM'...
Submodule 'include/boost/compute' (https://github.com/boostorg/compute) registered for path 'compute'
Submodule 'eigen' (https://gitlab.com/libeigen/eigen.git) registered for path 'eigen'
Submodule 'external_libs/fast_double_parser' (https://github.com/lemire/fast_double_parser.git) registered for path 'external_libs/fast_double_parser'
Submodule 'external_libs/fmt' (https://github.com/fmtlib/fmt.git) registered for path 'external_libs/fmt'
Cloning into '/LightGBM/compute'...
Cloning into '/LightGBM/eigen'...
Cloning into '/LightGBM/external_libs/fast_double_parser'...
Cloning into '/LightGBM/external_libs/fmt'...
Submodule path 'compute': checked out '36c89134d4013b2e5e45bc55656a18bd6141995a'
Submodule path 'eigen': checked out '8ba1b0f41a7950dc3e1d4ed75859e36c73311235'
Submodule path 'external_libs/fast_double_parser': checked out 'ace60646c02dc54c57f19d644e49a61e7e7758ec'
Submodule 'benchmark/dependencies/abseil-cpp' (https://github.com/abseil/abseil-cpp.git) registered for path 'external_libs/fast_double_parser/benchmarks/dependencies/abseil-cpp'
Submodule 'benchmark/dependencies/double-conversion' (https://github.com/google/double-conversion.git) registered for path 'external_libs/fast_double_parser/benchmarks/dependencies/double-conversion'
Cloning into '/LightGBM/external_libs/fast_double_parser/benchmarks/dependencies/abseil-cpp'...
Cloning into '/LightGBM/external_libs/fast_double_parser/benchmarks/dependencies/double-conversion'...
Submodule path 'external_libs/fast_double_parser/benchmarks/dependencies/abseil-cpp': checked out 'd936052d32a5b7ca08b0199a6724724aea432309'
Submodule path 'external_libs/fast_double_parser/benchmarks/dependencies/double-conversion': checked out 'f4cb2384efa55dee0e6652f8674b05763441ab09'
Submodule path 'external_libs/fmt': checked out 'cc09f1a6798c085c325569ef466bcdcffdc266d4'
* checking for file '/LightGBM/lightgbm_r/DESCRIPTION' ... OK
* preparing 'lightgbm':
* checking DESCRIPTION meta-information ... OK
* cleaning src
Warning in system2(command, args, stdout = NULL, stderr = NULL, ...) :
  error in running command
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
WARNING: directory 'lightgbm/src/compute/test' is empty
* looking to see if a 'data/datalist' file should be added
* building 'lightgbm_3.1.1.99.tar.gz'

* installing to library '/usr/local/lib/R/site-library'
* installing *source* package 'lightgbm' ...
** using staged installation
** libs
installing via 'install.libs.R' to /usr/local/lib/R/site-library/00LOCK-lightgbm/00new/lightgbm
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- R version passed into FindLibR.cmake: 4.0.3
-- Found LibR: /usr/lib/R
-- LIBR_EXECUTABLE: /usr/bin/R
-- LIBR_INCLUDE_DIRS: /usr/share/R/include
-- LIBR_CORE_LIBRARY: /usr/lib/R/lib/libR.so
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Looking for CL_VERSION_2_2
-- Looking for CL_VERSION_2_2 - found
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find OpenCL (missing: OpenCL_LIBRARY) (found version "2.2")
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindOpenCL.cmake:150 (find_package_handle_standard_args)
  CMakeLists.txt:138 (find_package)


-- Configuring incomplete, errors occurred!
See also "/tmp/RtmpvcXiAX/R.INSTALL14755eba078/lightgbm/src/build/CMakeFiles/CMakeOutput.log".
Error in .run_shell_command("cmake", c(cmake_args, "..")) :
  Command failed with exit code: 1
* removing '/usr/local/lib/R/site-library/lightgbm'
Error in .run_shell_command(install_cmd, install_args) :
  Command failed with exit code: 1
Execution halted
The command '/bin/sh -c git clone --recursive https://github.com/microsoft/LightGBM &&     cd LightGBM &&     Rscript build_r.R --use-gpu' returned a non-zero code: 1
```

If I build the docker image with the last `RUN` entry commented out:

```
FROM nvidia/cuda:11.0-devel-ubuntu20.04

RUN apt-get update && \
    DEBIAN_FRONTEND="noninteractive" apt-get install -y software-properties-common apt-transport-https

RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9 && \
    add-apt-repository 'deb [arch=amd64] https://cran.rstudio.com/bin/linux/ubuntu focal-cran40/' && \
    apt-get update && \
    apt-get install -y r-base

RUN apt-get install -y git wget libcurl4-openssl-dev default-jdk-headless libssl-dev libxml2-dev cmake

ENV MAKE="make -j$(nproc)"

RUN R -e 'install.packages(c("R6","data.table","jsonlite"), repos = "https://cran.rstudio.com/")'

RUN apt-get install -y libboost-dev libboost-system-dev libboost-filesystem-dev ocl-icd-opencl-dev opencl-headers clinfo

RUN mkdir -p /etc/OpenCL/vendors && \
    echo "libnvidia-opencl.so.1" > /etc/OpenCL/vendors/nvidia.icd   ## otherwise lightgm segfaults at runtime (compiles fine without it)

#RUN git clone --recursive https://github.com/microsoft/LightGBM && \
#    cd LightGBM && \
#    Rscript build_r.R --use-gpu
```

with

```
sudo docker build -t gbmperf_gpu .
```

and then run it:

```
sudo nvidia-docker run --rm -ti gbmperf_gpu /bin/bash
```

then I can run things manually:

```
git clone --recursive https://github.com/microsoft/LightGBM && \
    cd LightGBM && \
    Rscript build_r.R --use-gpu
```

gives the same error.

However, just compiling lightgbm (not the R package) seems fine:

```
git clone --recursive https://github.com/microsoft/LightGBM  && \
cd LightGBM  &&  mkdir build  &&  cd build  &&  cmake -DUSE_GPU=1 ..  &&  make -j4
```

as here:

```
...
Submodule path 'external_libs/fast_double_parser/benchmarks/dependencies/abseil-cpp': checked out 'd936052d32a5b7ca08b0199a6724724aea432309'
Submodule path 'external_libs/fast_double_parser/benchmarks/dependencies/double-conversion': checked out 'f4cb2384efa55dee0e6652f8674b05763441ab09'
Submodule path 'external_libs/fmt': checked out 'cc09f1a6798c085c325569ef466bcdcffdc266d4'
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Looking for CL_VERSION_2_2
-- Looking for CL_VERSION_2_2 - found
-- Found OpenCL: /usr/lib/x86_64-linux-gnu/libOpenCL.so (found version "2.2")
-- OpenCL include directory: /usr/include
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.71.0/BoostConfig.cmake (found suitable version "1.71.0", minimum required is "1.56.0") found components: filesystem system
-- Performing Test MM_PREFETCH
-- Performing Test MM_PREFETCH - Success
-- Using _mm_prefetch
-- Performing Test MM_MALLOC
-- Performing Test MM_MALLOC - Success
-- Using _mm_malloc
-- Configuring done
-- Generating done
-- Build files have been written to: /LightGBM/LightGBM/build
make[1]: warning: -j0 forced in submake: resetting jobserver mode.
Scanning dependencies of target lightgbm
Scanning dependencies of target _lightgbm
[  1%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/boosting.cpp.o
[  2%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt.cpp.o
[  4%] Building CXX object CMakeFiles/lightgbm.dir/src/boosting/gbdt.cpp.o
[  7%] Building CXX object CMakeFiles/lightgbm.dir/src/boosting/boosting.cpp.o
[  7%] Building CXX object CMakeFiles/lightgbm.dir/src/main.cpp.o
[  8%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/prediction_early_stop.cpp.o
[ 10%] Building CXX object CMakeFiles/lightgbm.dir/src/application/application.cpp.o
...
```

though I also see
```
/usr/include/CL/cl_version.h:34:104: note: #pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
   34 | #pragma message("cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)")
```

but it compiles anyway:

```
[ 98%] Linking CXX shared library ../lib_lightgbm.so
[100%] Linking CXX executable ../lightgbm
[100%] Built target _lightgbm
[100%] Built target lightgbm
```

So there must be something in the R package(?) cc @jameslamb

### PR DESCRIPTION
**Pull request by jameslamb**
This PR offers a fix for #3765. In that issue, @szilard described some issues using CMake-based builds of the LightGBM R package with GPU support. Specifically, compilation failed because LightGBM couldn't find OpenCL and Boost.

### Changes in this PR

This PR adds the following command-line options to `build_r.R`

* `--boost-root`
* `--boost-dir`
* `--boost-include-dir`
* `--boost-librarydir`
* `--opencl-include-dir`
* `--opencl-library`

The approach it takes is similar to how the Python package handles these same arguments:

* https://github.com/microsoft/LightGBM/blob/706f2af7badc26f6ec68729469ec6ec79a66d802/python-package/setup.py#L131-L144
* https://github.com/microsoft/LightGBM/blame/706f2af7badc26f6ec68729469ec6ec79a66d802/python-package/README.rst#L95-L111

### How I tested this

I adopted @szilard 's reproducible example from #3765 

1. Create a new directory and clone LightGBM into it

```shell
mkdir lgb-gpu-test
cd lgb-gpu-test
git clone --recursive https://github.com/microsoft/LightGBM.git
pushd LightGBM
    git fetch origin fix/r-gpu
    git checkout fix/r-gpu
popd
```

2. Write a file `Dockerfile` with the following content. The laptop I do GPU development on has CUDA 10.2 installed so I chose a CUDA 10.2 image, but I expect this would work for other versions.

<details><summary>Dockerfile (click me)</summary>

```dockerfile
FROM nvidia/cuda:10.2-devel-ubuntu18.04

ENV DEBIAN_FRONTEND="noninteractive"

RUN apt-get update && \
    apt-get install \
    	-y \
    	software-properties-common \
    	apt-transport-https

RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9 && \
    add-apt-repository 'deb [arch=amd64] https://cran.rstudio.com/bin/linux/ubuntu bionic-cran40/' && \
    apt-get update && \
    apt-get install -y \
    		r-base

RUN apt-get install -y \
	git \
	wget \
	libcurl4-openssl-dev \
	default-jdk-headless \
	libssl-dev \
	libxml2-dev \
	cmake

ENV MAKE="make -j$(nproc)"

RUN R -e 'install.packages(c("R6","data.table","jsonlite"), repos = "https://cran.rstudio.com/")'

RUN apt-get install -y \
		libboost-dev \
		libboost-system-dev \
		libboost-filesystem-dev \
		ocl-icd-opencl-dev \
		opencl-headers \
		clinfo

RUN mkdir -p /etc/OpenCL/vendors && \
    echo "libnvidia-opencl.so.1" > /etc/OpenCL/vendors/nvidia.icd   ## otherwise lightgm segfaults at runtime (compiles fine without it)

COPY LightGBM /tmp/LightGBM

RUN cd /tmp/LightGBM && \
	git submodule init && \
	git submodule update --recursive && \
    Rscript build_r.R \
    	--use-gpu \
    	--opencl-library=/usr/lib/x86_64-linux-gnu/libOpenCL.so \
    	--boost-librarydir=/usr/lib/x86_64-linux-gnu
```

</details>

3. Build the image

```shell
docker build --no-cache -t test-lgb-gpu -f Dockerfile .
```

4. Run a container from that image. Open two terminals. From one, open an R session in the container. In the other, open a shell running `nvidia-smi`.

```shell
nvidia-docker run \
	-w /tmp/LightGBM \
	-it test-lgb-gpu \
	R
```

```shell
nvidia-docker run \
	-w /tmp/LightGBM \
	-it gbmperf-lgb-gpu \
	watch -n 3 nvidia-smi
```

5. In the R shell, install some packages and then run @szilard 's test script from https://github.com/microsoft/LightGBM/issues/3765#issuecomment-760975539

<details><summary>test R script (click me)</summary>

```r
install.packages(c("ROCR", "curl"), repos = "https://cran.r-project.org")

library(data.table)
library(ROCR)
library(lightgbm)
library(Matrix)

set.seed(123)

d_train <- fread("https://s3.amazonaws.com/benchm-ml--main/train-1m.csv", showProgress=FALSE)
d_test <- fread("https://s3.amazonaws.com/benchm-ml--main/test.csv", showProgress=FALSE)

d_all <- rbind(d_train, d_test)
d_all$dep_delayed_15min <- ifelse(d_all$dep_delayed_15min=="Y",1,0)

d_all_wrules <- lgb.convert_with_rules(d_all)       
d_all <- d_all_wrules$data
cols_cats <- names(d_all_wrules$rules) 

d_train <- d_all[1:nrow(d_train)]
d_test <- d_all[(nrow(d_train)+1):(nrow(d_train)+nrow(d_test))]

p <- ncol(d_all)-1
dlgb_train <- lgb.Dataset(data = as.matrix(d_train[,1:p]), label = d_train$dep_delayed_15min, free_raw_data = FALSE)

md <- lgb.train(
	data = dlgb_train, 
	objective = "binary", 
	nrounds = 100, num_leaves = 512, learning_rate = 0.1, 
	categorical_feature = cols_cats,
	device = "gpu",
	verbose = 0
)

phat <- predict(md, data = as.matrix(d_test[,1:p]))
rocr_pred <- prediction(phat, d_test$dep_delayed_15min)
cat(performance(rocr_pred, "auc")@y.values[[1]],"\n")
```

</details>

Based on the output of `nvidia-smi`, I'm fairly sure that training is actually taking advantage of the GPU.

image

### Notes for Reviewers

* I think we should have a CI job for R + GPU, but that that's outside the scope of this PR. ~I'll add an issue and update here.~ #3780
* I chose not to hard-code any default values into `build_r.R`. I think this will make this more stable, even if it means users need to do a little bit more configuration. Let me know if you disagree with this.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
@jameslamb Please note this my comment in #3765
> I'm not sure that newer version from Ubuntu ppa is better than preinstalled native version from NVIDIA in case you are really using NVIDIA cards for training.
  https://github.com/microsoft/LightGBM/issues/3765#issuecomment-760933658

So I believe it worth to run tests **without** 

```
RUN apt-get install -y \
                ...
		ocl-icd-opencl-dev \
		opencl-headers \
                ...
```
before merging this PR. Could you please do this as I guess you already have easy access to the environment you've described in  your starting comment?

**Comment by jameslamb** [comment]
> @jameslamb Please note this my comment in #3765
> 
> > I'm not sure that newer version from Ubuntu ppa is better than preinstalled native version from NVIDIA in case you are really using NVIDIA cards for training.
> > [#3765 (comment)](https://github.com/microsoft/LightGBM/issues/3765#issuecomment-760933658)
> 
> So I believe it worth to run tests **without**
> 
> ```
> RUN apt-get install -y \
>                 ...
> 		ocl-icd-opencl-dev \
> 		opencl-headers \
>                 ...
> ```
> 
> before merging this PR. Could you please do this as I guess you already have easy access to the environment you've described in your starting comment?

I just tried this, and compilation failed

```text
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- R version passed into FindLibR.cmake: 4.0.3
-- Found LibR: /usr/lib/R  
-- LIBR_EXECUTABLE: /usr/bin/R
-- LIBR_INCLUDE_DIRS: /usr/share/R/include
-- LIBR_CORE_LIBRARY: /usr/lib/R/lib/libR.so
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Looking for CL_VERSION_2_2
-- Looking for CL_VERSION_2_2 - not found
-- Looking for CL_VERSION_2_1
-- Looking for CL_VERSION_2_1 - not found
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - not found
-- Looking for CL_VERSION_1_1
-- Looking for CL_VERSION_1_1 - not found
-- Looking for CL_VERSION_1_0
-- Looking for CL_VERSION_1_0 - not found
CMake Error at /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenCL (missing: OpenCL_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.10/Modules/FindOpenCL.cmake:132 (find_package_handle_standard_args)
  CMakeLists.txt:138 (find_package)


-- Configuring incomplete, errors occurred!
See also "/tmp/RtmpH3RoRU/R.INSTALL9e1d4796b2/lightgbm/src/build/CMakeFiles/CMakeOutput.log".
See also "/tmp/RtmpH3RoRU/R.INSTALL9e1d4796b2/lightgbm/src/build/CMakeFiles/CMakeError.log".
Error in .run_shell_command("cmake", c(cmake_args, "..")) : 
  Command failed with exit code: 1
* removing '/usr/local/lib/R/site-library/lightgbm'
Error in .run_shell_command(install_cmd, install_args) : 
  Command failed with exit code: 1
Execution halted
```

**Comment by StrikerRUS** [comment]
> I just tried this, and compilation failed

OK, expected.

Please try passing `-DOpenCL_LIBRARY=/usr/local/cuda/lib64/libOpenCL.so -DOpenCL_INCLUDE_DIR=/usr/local/cuda/include/` as script arguments, to make sure that this use case is also covered with new arguments.

**Comment by jameslamb** [comment]
Building like this, with `ocl-icd-opencl-dev` and `opencl-headers` installation removed,  succeeded :tada: 

```shell
Rscript build_r.R \
    	--use-gpu \
    	--opencl-library=/usr/local/cuda/lib64/libOpenCL.so \
    	--opencl-include-dir=/usr/local/cuda/include/
```

<details><summary>full Dockerfile</summary>

```dockerfile
FROM nvidia/cuda:10.2-devel-ubuntu18.04

ENV DEBIAN_FRONTEND="noninteractive"

RUN apt-get update && \
    apt-get install \
    	-y \
    	software-properties-common \
    	apt-transport-https

RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9 && \
    add-apt-repository 'deb [arch=amd64] https://cran.rstudio.com/bin/linux/ubuntu bionic-cran40/' && \
    apt-get update && \
    apt-get install -y \
    		r-base

RUN apt-get install -y \
	git \
	wget \
	libcurl4-openssl-dev \
	default-jdk-headless \
	libssl-dev \
	libxml2-dev \
	cmake

ENV MAKE="make -j$(nproc)"

RUN R -e 'install.packages(c("R6","data.table","jsonlite"), repos = "https://cran.rstudio.com/")'

RUN apt-get install -y \
		libboost-dev \
		libboost-system-dev \
		libboost-filesystem-dev \
		clinfo

RUN mkdir -p /etc/OpenCL/vendors && \
    echo "libnvidia-opencl.so.1" > /etc/OpenCL/vendors/nvidia.icd   ## otherwise lightgm segfaults at runtime (compiles fine without it)

COPY LightGBM /tmp/LightGBM

RUN cd /tmp/LightGBM && \
	git submodule init && \
	git submodule update --recursive && \
    Rscript build_r.R \
    	--use-gpu \
    	--opencl-library=/usr/local/cuda/lib64/libOpenCL.so \
    	--opencl-include-dir=/usr/local/cuda/include/

```

</details>

<details><summary>Build Logs (click me)</summary>

```text
-- Found OpenMP: TRUE (found version "4.5")  
-- Looking for CL_VERSION_2_2
-- Looking for CL_VERSION_2_2 - not found
-- Looking for CL_VERSION_2_1
-- Looking for CL_VERSION_2_1 - not found
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - found
-- Found OpenCL: /usr/local/cuda/lib64/libOpenCL.so (found version "1.2") 
-- OpenCL include directory: /usr/local/cuda/include
-- Boost version: 1.65.1
-- Found the following Boost libraries:
--   filesystem
--   system
-- Performing Test MM_PREFETCH
-- Performing Test MM_PREFETCH - Success
-- Using _mm_prefetch
-- Performing Test MM_MALLOC
-- Performing Test MM_MALLOC - Success
-- Using _mm_malloc
-- Configuring done
-- Generating done
...
[100%] Built target _lightgbm
Found library file: /tmp/RtmpaLCutl/R.INSTALL9e5da08f48/lightgbm/src/lib_lightgbm.so to move to /usr/local/lib/R/site-library/00LOCK-lightgbm/00new/lightgbm/libs
Removing 'build/' directory
** R
** data
** demo
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (lightgbm)
```
</details>

I ran the testing code but with `verbose = 1`, and can see the following logs confirming that the GPU was utilized

```text
[LightGBM] [Info] Number of positive: 192982, number of negative: 807018
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 1095
[LightGBM] [Info] Number of data points in the train set: 1000000, number of used features: 8
[LightGBM] [Info] Using GPU Device: GeForce RTX 2070 with Max-Q Design, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 8 dense feature groups (7.63 MB) transferred to GPU in 0.020953 secs. 0 sparse feature groups
```

File: build_r.R [hunk 1]
```diff
@@ -10,17 +10,58 @@ INSTALL_AFTER_BUILD <- !("--skip-install" %in% args)
...
+parsed_args <- .parse_args(args)
+
+USING_GPU <- "--use-gpu" %in% parsed_args[["flags"]]
+USING_MINGW <- "--use-mingw" %in% parsed_args[["flags"]]
+USING_MSYS2 <- "--use-msys2" %in% parsed_args[["flags"]]
+
+# this maps command-line arguments to defines passed into CMake,
+ARGS_TO_DEFINES <- c(
+  "--boost-root" = "-DBOOST_ROOT"
+  , "--boost-dir" = "-Dboost_DIR"
```
Comment (StrikerRUS) [inline build_r.R:46]:
```suggestion
  , "--boost-dir" = "-DBoost_DIR"
```

**Review by StrikerRUS — APPROVED** [review]
LGTM except one typo! But I haven't dug deep into R code.

**Comment by jameslamb** [comment]
> ... I haven't dug deep into R code.

I think that since we have so many CI jobs for `R + CMake`, and since this PR didn't change any CI scripts or tests, we can be pretty confident that the changes to `build_r.R` didn't break the experience of building the CPU package with `Rscript build_r.R`. All of the non-GPU command line options are covered by tests

* `Rscript build_r.R --use-mingw`: https://github.com/microsoft/LightGBM/blob/c871496a3ecac2101dae3f696096198feab3f281/.ci/test_r_package_windows.ps1#L130
* `Rscript build_r.R --use-msys2`: https://github.com/microsoft/LightGBM/blob/c871496a3ecac2101dae3f696096198feab3f281/.ci/test_r_package_windows.ps1#L133
* `Rscript build_r.R --skip-install`
    - https://github.com/microsoft/LightGBM/blob/c871496a3ecac2101dae3f696096198feab3f281/.ci/test_r_package_windows.ps1#L130-L135
    - https://github.com/microsoft/LightGBM/blob/c871496a3ecac2101dae3f696096198feab3f281/.ci/test_r_package.sh#L113

**Pull request merged**